File: api/src/main/java/org/vosao/common/Messages.java
/**
 * Vosao CMS. Simple CMS for Google App Engine.
 * Copyright (C) 2009 Vosao development team
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * email: vosao.dev@gmail.com
 */

package org.vosao.common;

import java.text.MessageFormat;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.ResourceBundle;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpSession;

/**
 * Message bundle helper class for creatng localized messages from Java code.
 * 
 * @author Alexander Oleynik
 *
 */
public class Messages {

	public static final String JSTL_FMT_LOCALE_KEY = 
		"javax.servlet.jsp.jstl.fmt.locale.session";

	private static final String BUNDLE_BASE_NAME = 
		"org.vosao.resources.messages";
	
	private static final Locale[] supportedLocales = {
		Locale.ENGLISH, 
		new Locale("ru")};
	
	private static ResourceBundle getBundle(HttpServletRequest request) {
		HttpSession session = request.getSession();
		Locale sessionLocale = (Locale)session.getAttribute(JSTL_FMT_LOCALE_KEY);
		Locale locale = sessionLocale != null ? sessionLocale : 
			request.getLocale();
		return ResourceBundle.getBundle(BUNDLE_BASE_NAME, locale);
	}
	
	private static ResourceBundle getDefaultBundle() {
		return ResourceBundle.getBundle(BUNDLE_BASE_NAME, Locale.ENGLISH);
	}

	public static String get(String key, Object...objects) {
		VosaoContext ctx = VosaoContext.getInstance();
		String pattern = "not found";
		if (isLocaleSupported(ctx.getLocale())) {
			pattern  = getBundle(ctx.getRequest()).getString(key);
		}
		else {
			pattern  = getDefaultBundle().getString(key);
		}
		if (objects != null) {
			MessageFormat formatter = new MessageFormat("");
			formatter.setLocale(ctx.getLocale());
			formatter.applyPattern(pattern);
			pattern = formatter.format(objects);
		}
		return pattern;
	}
	
	/**
	 * Generate JavaScript JSON message bundle for JavaScript messages 
	 * localization.
	 * @return JS file.
	 */
	public static String getJSMessages() {
		VosaoContext ctx = VosaoContext.getInstance();
		Map<String, String> messages = new HashMap<String, String>();
		ResourceBundle defaultBundle = getDefaultBundle();
		for (String key : defaultBundle.keySet()) {
			messages.put(key, defaultBundle.getString(key));
		}
		ResourceBundle bundle = getBundle(ctx.getRequest());
		for (String key : bundle.keySet()) {
			messages.put(key, bundle.getString(key));
		}
		StringBuffer result = new StringBuffer();
		result.append("locale = '").append(ctx.getLocale().toString())
				.append("';\n");
		result.append("messages = {\n");
		int i = 0;
		for (String key : messages.keySet()) {
			if (i++ > 0) {
				result.append(",");
			}
			result.append(" '").append(key).append("' : '")
				.append(filterForJS(messages.get(key)))
				.append("'\n");
		}
		result.append("};");
		return result.toString();
	}
	
	private static String filterForJS(String msg) {
		return msg.replaceAll("\n", "").replaceAll("'", "\\\\'");
	}
	
	public static boolean isLocaleSupported(Locale locale) {
		for (Locale l : supportedLocales) {
			if (l.equals(locale)) {
				return true;
			}
		}
		return false;
	}

}

File: branches/single/web/src/main/webapp/static/js/cms/config/messages.js
/**
 * Vosao CMS. Simple CMS for Google App Engine.
 * 
 * Copyright (C) 2009-2010 Vosao development team.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * email: vosao.dev@gmail.com
 */

var languages = null;

$(function(){
    $("#message-dialog").dialog({ width: 400, autoOpen: false });
    Vosao.initJSONRpc(loadData);
    $('#addMessageButton').click(onAddMessage);
    $('#removeMessageButton').click(onRemoveMessage);
    $('#messageForm').submit(function() {onMessageSave(); return false;});
    $('#cancelMessageDlgButton').click(onMessageCancel);
    $('ul.ui-tabs-nav li:nth-child(4)').addClass('ui-state-active')
    		.addClass('ui-tabs-selected')
			.removeClass('ui-state-default');
});

function loadData() {
	loadLanguages();
}

// Language

function loadLanguages() {
	Vosao.jsonrpc.languageService.select(function (r) {
        languages = r.list;
        loadMessages();
    });
}

// Message 

function onAddMessage() {
	clearDialogMessages();
    var h = '';
	$.each(languages, function (i, lang) {
        h += '<div class="form-row"><label>' + lang.title 
            + '</label><input type="text" id="message_' + lang.code + '" /></div>';
    });
    $('#messageCode').val('');
	$('#messagesInput').html(h);
	$('#message-dialog').dialog('open');
}

function onRemoveMessage() {
    var codes = [];
    $('#messageBundle input:checked').each(function () {
        codes.push(this.value);
    });
    if (codes.length == 0) {
    	Vosao.info(messages('nothing_selected'));
        return;
    }
    if (confirm(messages('are_you_sure'))) {
    	Vosao.jsonrpc.messageService.remove(function (r) {
    		Vosao.info(r.message);
            loadMessages();
        }, Vosao.javaList(codes));
    }
}

function loadMessages() {
	Vosao.jsonrpc.messageService.select(function (r) {
        var h = '<table class="form-table"><tr><th></th><th>' + messages('code') 
        	+ '</th>';
        $.each(languages, function (i, lang) {
            h += '<th>' + lang.title + '</th>';
        });
        h += '</tr>';
        $.each(r.list, function (i, msg) {
            h += '<tr><td><input type="checkbox" value="' + msg.code + '"/></td>\
                <td><a href="#" onclick="onMessageEdit(\'' + msg.code + '\')">'
                + msg.code + '</a></td>';
            $.each(languages, function (i, lang) {
                h += '<td>' + msg.values.map[lang.code] + '</td>';
            });
            h += '</tr>';
        });
        $('#messageBundle').html(h + '</table>');
        $('#messageBundle tr:even').addClass('even');
	});
}

function onMessageEdit(code) {
	Vosao.jsonrpc.messageService.selectByCode(function (r) {
        onAddMessage();
        $('#messageCode').val(code);
        $.each(r.list, function (i, msg) {
            $('#message_' + msg.languageCode).val(msg.value);
        });
    }, code);
}

function onMessageSave() {
    var vo = {code : $('#messageCode').val() };
    $.each(languages, function (i, lang) {
        vo[lang.code] = $('#message_' + lang.code).val();
    });
    Vosao.jsonrpc.messageService.save(function (r) {
        if (r.result == 'success') {
            loadMessages();
            $('#message-dialog').dialog('close');
        }
        else {
            messageErrors(r.messages.list);
        }
    }, Vosao.javaMap(vo));
    
}

function onMessageCancel() {
    $('#message-dialog').dialog('close');
}

function messageError(msg) {
	Vosao.errorMessage('#message-dialog .messages', msg);
}

function messageErrors(errors) {
	Vosao.errorMessages('#message-dialog .messages', errors);
}

function clearDialogMessages() {
	$('#message-dialog .messages').html();
}
